From 66140b0c926ed097f2cb7474863523e4af351f5b Mon Sep 17 00:00:00 2001
From: Rich Felker <dalias@aerifal.cx>
Date: Tue, 2 Dec 2014 22:17:52 -0500
Subject: [PATCH] fix return value of pthread_getaffinity_np and
 pthread_setaffinity_np

these functions are expected to return an error code rather than
setting errno and returning -1.
---
 src/sched/affinity.c | 19 +++++++++++--------
 1 file changed, 11 insertions(+), 8 deletions(-)

diff --git a/src/sched/affinity.c b/src/sched/affinity.c
index 737e41b..948ece4 100644
--- a/src/sched/affinity.c
+++ b/src/sched/affinity.c
@@ -11,20 +11,23 @@ int sched_setaffinity(pid_t tid, size_t size, const cpu_set_t *set)
 
 int pthread_setaffinity_np(pthread_t td, size_t size, const cpu_set_t *set)
 {
-	return syscall(SYS_sched_setaffinity, td->tid, size, set);
+	return -__syscall(SYS_sched_setaffinity, td->tid, size, set);
 }
 
-int sched_getaffinity(pid_t tid, size_t size, cpu_set_t *set)
+static int do_getaffinity(pid_t tid, size_t size, cpu_set_t *set)
 {
 	long ret = __syscall(SYS_sched_getaffinity, tid, size, set);
-	if (ret > 0) {
-		if (ret < size) memset((char *)set+ret, 0, size-ret);
-		ret = 0;
-	}
-	return __syscall_ret(ret);
+	if (ret < 0) return ret;
+	if (ret < size) memset((char *)set+ret, 0, size-ret);
+	return 0;
+}
+
+int sched_getaffinity(pid_t tid, size_t size, cpu_set_t *set)
+{
+	return __syscall_ret(do_getaffinity(tid, size, set));
 }
 
 int pthread_getaffinity_np(pthread_t td, size_t size, cpu_set_t *set)
 {
-	return sched_getaffinity(td->tid, size, set);
+	return -do_getaffinity(td->tid, size, set);
 }
-- 
2.2.0

